tests/unit_tests/dft/include/reference_dft.hpp

```diff
@@ -0,0 +1,49 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+#ifndef ONEMKL_REFERENCE_DFT_HPP
+#define ONEMKL_REFERENCE_DFT_HPP
+
+#include <cmath>
+#include <complex>
+#include <vector>
+
+#include "test_common.hpp"
+
+template <typename TypeIn, typename TypeOut>
+void reference_forward_dft(const std::vector<TypeIn> &in, std::vector<TypeOut> &out) noexcept {
+    using ref_t = long double; /* Do the calculations using long double */
+    static_assert(is_complex<TypeOut>(), "Output type of DFT must be complex");
+
+    const ref_t TWOPI = ref_t{ 2.0 } * std::atan(ref_t{ 1.0 }) * ref_t{ 4.0 };
+
+    const size_t N = out.size();
+    for (std::size_t k = 0; k < N; ++k) {
+        std::complex<ref_t> out_temp = 0;
+        const auto partial_expo = (static_cast<ref_t>(k) * TWOPI) / static_cast<ref_t>(N);
+        for (std::size_t n = 0; n < N; ++n) {
```

Comment:
I think the design here is for this reference computation to not distinguish between real and complex input and always compute the full length output. There's a cast of the input to complex type below and some code in the compute_*.hpp files to construct a reference output of length `conjugate_even_size` from this initial reference output in the case of real.

Reply:
To me that over-allocation just looks like a design decision to allocate a full `in.size()` complex numbers for the reference output and compute the full output in the ref implementation, presumably so the reference doesn't have to have any conditionals on input type. I see your point about it being potentially confusing, but I don't see anything unsafe.

Reply:
There is nothing "unsafe" in the current implementation (wrong wording from myself), but my opinion was that this function is susceptible to produce an incorrect reference solution in case of slightly modified usage although such usage would seem acceptable from the signature alone. The complex roots of unity are wrong if `out.size()` is not equal to the length of the transform being tested. 

This usage assumption is not broken anywhere in the current usage/PR but it's not guarded anywhere, either. I don't think it would be detrimental in any way to have the length of the DFT under test an extra argument to this function, say `dft_length` (unless we no longer used `N = out.size()`), and have 
`const auto partial_expo = (static_cast<ref_t>(k) * TWOPI) / static_cast<ref_t>(dft_length);`
and 
`for (std::size_t n = 0; n < dft_length; ++n)`
in the body of this reference implementation, with another appropriate guard on `in.size() >= dft_length` herein. That would make it valid in the current usage but also in possible other use cases as I was foreseeing when I started this conversation, unless I'm mistaken.

This being said, the descriptive comment added in [774e322901b78a5bee2f1021f6e120559728b49d](https://github.com/oneapi-src/oneMKL/pull/261/commits/774e322901b78a5bee2f1021f6e120559728b49d) alleviates some of my concern.

Reply:
Thanks. I certainly am not opposed to any of those suggestions.

Reply:
As I understand it, the concern is that there is scope to compute an incorrect result when `in.size() != out.size()`. I've added an exception for this in [3cd6090](https://github.com/oneapi-src/oneMKL/pull/261/commits/3cd6090814dd7e81abe88942fefd7f07c30df34b). This function is only used in the DFT_test init, so hopefully there is limited scope for misuse.

I think this function gains more significant changes when multi-dimensional DFT tests are implemented in a later PR.